File: apps/app/lib/features/checkout/domain/entities/coupon.dart
import 'package:equatable/equatable.dart';

/// A coupon document (`coupons/{id}`), applied at checkout.
class Coupon extends Equatable {
  final String id;
  final String code;
  final String description;

  /// 'Percentage' | 'Fixed'.
  final String discountType;
  final double discountValue;
  final double minimumOrderAmount;
  final double maximumDiscount;
  final int usageLimit;
  final int usedCount;
  final int perUserLimit;
  final DateTime? validFrom;
  final DateTime? validUntil;
  final bool isActive;
  final List<String> applicableCategories;
  final List<String> applicableProducts;

  /// 'All' | 'Category' | 'Product'.
  final String applicationType;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Coupon({
    required this.id,
    required this.code,
    required this.description,
    required this.discountType,
    required this.discountValue,
    required this.minimumOrderAmount,
    required this.maximumDiscount,
    required this.usageLimit,
    required this.usedCount,
    required this.perUserLimit,
    required this.validFrom,
    required this.validUntil,
    required this.isActive,
    required this.applicableCategories,
    required this.applicableProducts,
    required this.applicationType,
    required this.createdAt,
    required this.updatedAt,
  });

  @override
  List<Object?> get props => [
        id,
        code,
        description,
        discountType,
        discountValue,
        minimumOrderAmount,
        maximumDiscount,
        usageLimit,
        usedCount,
        perUserLimit,
        validFrom,
        validUntil,
        isActive,
        applicableCategories,
        applicableProducts,
        applicationType,
        createdAt,
        updatedAt,
      ];
}

File: packages/shared/src/types/coupon.ts
import { Timestamp } from 'firebase/firestore';

/**
 * CouponProps — the shape of a coupon document in the `coupons` collection (admin Feature 12 / spec §1.12).
 * WHY in packages/shared: the coupon contract is consumed by BOTH the admin panel (CRUD) and the
 * customer app (redemption at checkout), so it is defined once here and imported by both — never
 * duplicated per app (skill: shared domain types live in packages/shared).
 * WHY Timestamp (not number) for date fields: preserves server-clock precision and sorts/range-queries
 * correctly server-side (skill: date/time fields are Firestore Timestamps).
 */
export interface CouponProps {
  id: string;
  code: string;
  description: string;
  discountType: 'Percentage' | 'Fixed';
  discountValue: number;
  minimumOrderAmount: number;
  maximumDiscount: number;
  usageLimit: number;
  usedCount: number;
  perUserLimit: number;
  validFrom: Timestamp;
  validUntil: Timestamp;
  isActive: boolean;
  applicableCategories: string[];
  applicableProducts: string[];
  applicationType: 'All' | 'Category' | 'Product';
  createdBy: string;
  createdByName: string;
  keywords: string[];
  createdAt: Timestamp;
  updatedAt: Timestamp;
}
